src/rdkafka_txnmgr.c

```diff
@@ -1192,8 +1194,13 @@ static rd_kafka_error_t *rd_kafka_txn_curr_api_req(rd_kafka_t *rk,
 
         rd_snprintf(rk->rk_eos.txn_curr_api.name,
                     sizeof(rk->rk_eos.txn_curr_api.name), "%s", name);
-
-        tmpq = rd_kafka_q_new(rk);
+        /* When API changes, destroy previous pending queue. */
```

Comment:
This is a nice solution, but do we really want to support API changes in the midst of an operation? It seems complex to recover cleanly from.

Say you have commit_transaction(timeout=300ms) and the API call times out while awaiting response from the broker, but the operation continues in the background.
Now the application calls abort_transaction(): what do we do here?
If we receive a succesful commit response then we should probably fail abort_transaction() - since the abort is no longer possible. But how do we make it meaningful for the application to recover from that?

I think it would be better to simply not allow other API calls if one has been started.

What do you think?

I have some work in progress for this that we can look at together.

Reply:
The intention of this change was to keep having one transactional call at a time, just to make some retriable calls like commit and abort able to retry when timed out without sending a new EndTxn. Could add a test case for that situation, but first let's talk about your work.

Reply:
Doing a quick check, if abort_transaction is called on the same thread after an initial retriable timeout on commit, it give a STATE error with message "Conflicting commit_transaction call already in progress". It's worth checking what happens from a different thread.